File: ts/isIsomorphic.ts
function isIsomorphic(s: string, t: string): boolean {
  const sMap = new Map<string, string>();
  const tMap = new Map<string, string>();

  for (let i = 0; i < s.length; i++) {
    const sEl = sMap.get(s[i]);
    const tEl = tMap.get(t[i]);

    if (!sEl && !tEl) {
      sMap.set(s[i], t[i]);
      tMap.set(t[i], s[i]);
      continue;
    }

    if (sEl !== t[i] || tEl !== s[i]) {
      return false;
    }
  }

  return true;
}

export default () => {
  // const s = 'badc';
  // const t = 'baba';
  // const s = 'foo';
  // const t = 'bar';
  // const s = 'paper';
  // const t = 'title';
  const s = 'egg';
  const t = 'add';

  const result = isIsomorphic(s, t);
  console.log(result);
};

File: rs/src/is_isomorphic.rs
use std::collections::{HashMap, HashSet};

fn is_isomorphic(s: String, t: String) -> bool {
    if s.len() != t.len() {
        return false;
    }

    let mut s_map: HashMap<char, char> = HashMap::new();
    let mut t_set: HashSet<char> = HashSet::new();

    for (s_char, t_char) in s.chars().zip(t.chars()) {
        if let Some(&mapped_char) = s_map.get(&s_char) {
            if mapped_char != t_char {
                return false;
            }
        } else {
            if t_set.contains(&t_char) {
                return false;
            }

            s_map.insert(s_char, t_char);
            t_set.insert(t_char);
        }
    }

    return true;
}

pub fn main() {
    let s = "egg".to_string();
    let t = "add".to_string();

    let result = is_isomorphic(s, t);
    println!("{}", result);
}
